PrestaShop/PrestaShop
# SpecificPrice priority form

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add specific price priority list handling in product page. It differs abit from old behavior. Now merchant can select to either have specific list for product, or use default settings (default settings added in configure -> shop parameters -> product settings ->general tab). Styling abit off in general settings, but doesn't block (its probably due to fact its using older theme and useless empty label divs are hanging there).
| Type?             | refacto
| Category?         | BO
| BC breaks?        | yes (see bellow)
| Deprecations?     | no
| Fixed ticket?     | #19674
| How to test?      | Check bellow.
| Possible impacts? | 

**Screenshots**
![Screenshot from 2022-01-31 16-12-10](https://user-images.githubusercontent.com/31609858/151808929-55ad7aa1-ea89-4b44-a3d5-8175a3ed2cde.png)

![Screenshot from 2022-01-25 20-35-27](https://user-images.githubusercontent.com/31609858/151043437-8afdd484-5f98-41ed-9f06-2f75daa92ca1.png)

**SPECS**

https://github.com/PrestaShop/prestashop-specs/blob/master/content/1.7/back-office/catalog/products/add-edit-v2.md#specific-prices

**BC BREAKS**
Following BC breaks only relates to experimental product page.
- `PrestaShop\PrestaShop\Core\Domain\Product\QueryResult\ProductPricesInformation` added argument `PriorityList|null $specificPricePriorities` in constructor
- `PrestaShop\PrestaShop\Adapter\Product\QueryHandler\GetProductForEditingHandler` added argument `SpecificPriceRepository $specificPricePriorities` in constructor
- `PrestaShop\PrestaShop\Core\Domain\Product\SpecificPrice\Exception\CannotSetSpecificPricePrioritiesException` removed as not used anymore (and it was recently inroduced with v1 of experimental product page)

**DEPRECATIONS**
* SpecificPrice::setPriorities() - use SpecificPricePriorityUpdater->setPriorities() instead. New method does not remove all specific prices.

**How to test**
1. Go to experimental product page (don't forget to enable experimental product page feature in "advanced settings -> experimental features")
2. In bottom of pricing tab you should see priorities list as displayed in screenshot.
3. When "Use default priority settings" is selected there should be no list of priorities.
4. When "Specific priority settings to this product" is selected there should appear a list of priorities.
5. Selecting same option in list multiple times is not allowed (it should throw general error at the top of screen on submit "The selected condition must be different in each field to set an order of priority.". UX should be improved within next pull requests).
4. "Manage default settings" link should lead to Shop parameters -> Product configuration settings.
5. In `Shop parameters -> Product configuration General tab` you should see the same priorities list. And it should also show error when multiple same options are selected.
6. All these priority settings are associated with Front office specific prices calculation. When product has specific list, those priorities should be applied for it. If product has no specific list (the option "Use default priority settings" is selected) then default priorities from general product configuration should be applied.
7. :warning: styling was not focused too much on this PR, because the visuals are expected to change (to enhance user experience).



---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/26938)
